src/compile_cache.cc

```diff
@@ -0,0 +1,378 @@
+// [uint32_t] cache hash
+// .... compile cache content ....
+void CompileCacheHandler::Persist() {
+  DCHECK(!compile_cache_dir_.empty());
+
+  // NOTE(joyeecheung): in most circumstances the code caching reading
+  // writing logic is lenient enough that it's fine even if someone
+  // overwrites the cache (that leads to either size or hash mismatch
+  // in subsequent loads and the overwritten cache will be ignored).
+  // Also in most use cases users should not change the files on disk
+  // too rapidly. Therefore locking is not currently implemented to
+  // avoid the cost.
+  for (auto& pair : compiler_cache_store_) {
+    auto* entry = pair.second.get();
+    if (entry->cache == nullptr) {
+      Debug("[compile cache] skip %s because the cache was not initialized\n",
+            entry->source_filename);
+      continue;
+    }
+    if (entry->refreshed == false) {
+      Debug("[compile cache] skip %s because cache was the same\n",
+            entry->source_filename);
+      continue;
+    }
+
+    DCHECK_EQ(entry->cache->buffer_policy,
+              v8::ScriptCompiler::CachedData::BufferOwned);
+    char* cache_ptr =
+        reinterpret_cast<char*>(const_cast<uint8_t*>(entry->cache->data));
+    uint32_t cache_size = static_cast<uint32_t>(entry->cache->length);
+    uint32_t cache_hash = GetHash(cache_ptr, cache_size);
+
+    // Generating headers.
+    std::vector<uint32_t> headers(kHeaderCount);
+    headers[kCodeSizeOffset] = entry->code_size;
+    headers[kCacheSizeOffset] = cache_size;
+    headers[kCodeHashOffset] = entry->code_hash;
+    headers[kCacheHashOffset] = cache_hash;
+
+    Debug("[compile cache] writing cache for %s in %s [%d %d %d %d]...",
+          entry->source_filename,
+          entry->cache_filename,
+          headers[kCodeSizeOffset],
+          headers[kCacheSizeOffset],
+          headers[kCodeHashOffset],
+          headers[kCacheHashOffset]);
+
+    uv_buf_t headers_buf = uv_buf_init(reinterpret_cast<char*>(headers.data()),
+                                       headers.size() * sizeof(uint32_t));
+    uv_buf_t data_buf = uv_buf_init(cache_ptr, entry->cache->length);
+    uv_buf_t bufs[] = {headers_buf, data_buf};
+
+    int err = WriteFileSync(entry->cache_filename.c_str(), bufs, 2);
+    if (err < 0) {
+      Debug("failed: %s\n", uv_strerror(err));
+    } else {
+      Debug("success\n");
+    }
+  }
+}
+
+CompileCacheHandler::CompileCacheHandler(Environment* env)
+    : isolate_(env->isolate()),
+      is_debug_(
+          env->enabled_debug_list()->enabled(DebugCategory::COMPILE_CACHE)) {}
+
+// Directory structure:
+// - Compile cache directory (from NODE_COMPILE_CACHE)
+//   - <cache_version_tag_1>: hash of CachedDataVersionTag + NODE_VERESION
+//   - <cache_version_tag_2>
+//   - <cache_version_tag_3>
+//     - <cache_file_1>: a hash of filename + module type
+//     - <cache_file_2>
+//     - <cache_file_3>
+bool CompileCacheHandler::InitializeDirectory(const std::string& dir) {
```

Comment:
(Not blocking for this PR - we should have better user facing errors if there are permission issues creating/reading/writing the cache)

Reply:
This should not error at all, the idea is that this will just silently ignore any errors if the cache cannot be done.

Reply:
It's literally caching I get that but when it won't work for users and they'll want to figure out why they'll ask us. It would be useful to at least trace it as as. warning or add DEBUG* like logs.

Reply:
There are already traces. The handful of tests here are testing against the traces .